conf/tls/tls_config.go

```diff
@@ -0,0 +1,125 @@
+}
+
+// Decode is the custom kong decoder that is similar to kong.NamedFileContentFlag's decode implementation,
+// but stores the error from ioutil.ReadFile instead of returning it.
+func (f *NamedFileContent) Decode(ctx *kong.DecodeContext) error { // nolint: golint
+	var filename string
+	err := ctx.Scan.PopValueInto("filename", &filename)
+	if err != nil {
+		return errors.Errorf("Failed to retrieve filename flag: %v", err)
+	}
+	// This allows unsetting of file content flags.
+	if filename == "" {
+		*f = NamedFileContent{}
+		return nil
+	}
+	filename = kong.ExpandPath(filename)
+	data, err := ioutil.ReadFile(filename) // nolint: gosec
+	if err != nil {
+		// Instead of returning an error, set contents to nil and store the error.
+		// The error will be checked during validation when building the configuration.
+		f.Contents = nil
+		f.Error = err
+	} else {
+		f.Contents = data
+	}
+
+	f.Filename = filename
+	return nil
+}
+
+// validateTLSFiles returns the first error it finds from the NamedFileContent fields. The errors are
+// populated by kong during the decode step, and are evaluated during validation.
+func (t *TLSConfig) validateTLSFiles() error {
+	if t.CACert.Error != nil {
+		return errors.Errorf("failed to open %q: %s", t.CACert.Filename, t.CACert.Error)
+	}
+
+	if t.Key.Error != nil {
+		return errors.Errorf("failed to open %q: %s", t.Key.Filename, t.Key.Error)
+	}
+
+	if t.Cert.Error != nil {
+		return errors.Errorf("failed to open %q: %s", t.Cert.Filename, t.Cert.Error)
+	}
+	return nil
+}
+
+// BuildTLSConfig loads CA certs and public/private key pair given options.
+//
+// Will return (nil, nil) if both cert and keyare not specified.
+func BuildTLSConfig(config TLSConfig) (*tls.Config, error) {
+	if err := config.validateTLSFiles(); err != nil {
+		return nil, err
+	}
+	if !config.IsConfigured() {
+		return nil, nil
+	}
+	tlsConfig := &tls.Config{ // nolint: gosec
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
+	}
+	if len(config.CACert.Contents) != 0 {
+		rootCertPool := x509.NewCertPool()
+		if ok := rootCertPool.AppendCertsFromPEM(config.CACert.Contents); !ok {
+			return nil, errors.Errorf("failed to append CA PEM from (invalid PEM block?)")
+		}
+		tlsConfig.RootCAs = rootCertPool
+		tlsConfig.ClientCAs = rootCertPool
+	}
+	if len(config.Key.Contents) != 0 && len(config.Cert.Contents) != 0 {
+		certs, err := tls.X509KeyPair(config.Cert.Contents, config.Key.Contents)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse client key pair")
+		}
+		tlsConfig.Certificates = []tls.Certificate{certs}
+	}
+	clientAuth, ok := clientAuthTypeMap[config.ClientAuth]
+	if !ok {
+		return nil, errors.Errorf("invalid tls client auth setting '%s'", config.ClientAuth)
+	}
+	tlsConfig.ClientAuth = clientAuth
+	tlsConfig.BuildNameToCertificate()
+
+	return tlsConfig, nil
+}
+
+// IsConfigured returns true if TLS is configured. ie. CA cert and keys are provided.
+func (t *TLSConfig) IsConfigured() bool {
```

Comment:
Not sure what config.IsConfigured() means. Isn't all config, by definition, configured?
Perhaps it should be called hasTLS()

Reply:
Also, I don't think it belongs here as only used when converting to another type of tls config required for grpc

Reply:
Removed